### PR TITLE
Modloader

### PR DESCRIPTION
**Pull request by DJ1TJOO**
The modloader is an addition to shapez.io for easier access to mods and it is more organized for the user. The modloader adds the shapezAPI and makes existing classes easier to edit. Ofcourse mods can still overwrite the existing functions, variables and classes if nessesary. 

### The modloader
**Added:**
- Choosing a gamemode
- ShapezAPI exports classes, fuctions and variables so mods can access them.
- Modsmanager to load mods.
- A new version because building codes where changed to strings
- Extra calls for mods when language changes and root is initializing

**The changes:**
- The registries for buildings, items, components, systems, gamemodes and states are now going through the shapezAPI and then register them. You can add a building, item, component, system, gamemode and/or state to the shapezAPI before registering and your building/item/component/system/gamemode/state is registered.
- All buildings have static variables with functions for different variants of the building. Because of this multiple mods can edit the same building but different variants.
- All componetens, systems, buildings and gamemodes have ids.
- Buildings codes are now strings with `buildingid/variant/rotationVariant`.
- A couple of components and systems have static variables to.
- MapView goes through every system and checks if it has the function for rendering. An order is specified in the shapezAPI.
- Ability to easily add hud parts

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -130,3 +130,8 @@ For most assets I use Adobe Photoshop, you can find them <a href="//github.com/t
 All assets will be automatically rebuilt into the atlas once changed (Thanks to dengr1065!)
 
 <img src="https://i.imgur.com/W25Fkl0.png" alt="shapez.io Screenshot">
+
+For generating docs:
+npm run esdoc
+Replace name.replace('\*', '\\\\\*') with name.replace('\*', '\\\\\*').replace('[', '\\\\[').replace(']', '\\\\]').replace('(', '\\\\(').replace(')', '\\\\)') in node_modules\esdoc-publish-html-plugin\out\src\Builder\DocBuilder.js:116:20)
```
Comment (tobspr) [inline README.md:136]: Maybe this could be formatted a bit more?

File: electron/index.js [hunk 2]
```diff
@@ -153,63 +154,147 @@ ipcMain.on("exit-app", (event, flag) => {
...
-                return {
-                    error: ex,
-                };
-            }
-
-            return {
-                success: true,
-                data: job.contents,
-            };
+    if (job.mods) {
```
Comment (tobspr) [inline electron/index.js:157]: I think it is unnecessary here to make it an if-else or? instead just compute the fname differently?

Comment (tobspr) [reply electron/index.js:157]: also, maybe we can somehow ensure that the game can not read any file outside of its game folder? to make sure mods can not access sensitive data

File: src/css/animations.scss [hunk 3]
```diff
@@ -1,19 +1,13 @@
...
+    }
 
-        100% {
-            transform: scale(1, 1);
-        }
+    50% {
+        transform: scale(1.03, 1.03);
     }
 
-    .#{$animName} {
```
Comment (tobspr) [inline src/css/animations.scss:16]: why did you remove it here?

File: src/css/dynamic_ui.scss [hunk 4]
```diff
@@ -22,22 +21,18 @@
 // Helper method to scale a property value
 @mixin S($propName, $v1, $v2: "", $v3: "", $v4: "", $important: false) {
     $impSuffix: "";
-    @if $important == true {
+    @if $important==true {
```
Comment (tobspr) [inline src/css/dynamic_ui.scss:24]: missing spaces

File: src/css/dynamic_ui.scss [hunk 5]
```diff
@@ -22,22 +21,18 @@
...
         $impSuffix: "!important";
     }
-
     $v1: D($v1);
```
Comment (tobspr) [inline src/css/dynamic_ui.scss:27]: somehow this file got autoformatted, too

File: src/js/application.js [hunk 6]
```diff
@@ -118,6 +111,7 @@ export class Application {
         // Store the mouse position, or null if not available
         /** @type {Vector|null} */
         this.mousePosition = null;
+        this.onAboutClicked = undefined;
```
Comment (tobspr) [inline src/js/application.js:114]: What is this doing here?

File: src/js/core/config.local [hunk 7]
```diff
@@ -0,0 +1,114 @@
+export default {
```
Comment (tobspr) [inline src/js/core/config.local:1]: Could you remove this file from the PR? It should get replaced soon anyways

File: src/js/core/read_write_proxy.js [hunk 8]
```diff
@@ -317,7 +318,7 @@ export class ReadWriteProxy {
         if (!data) {
             return ExplainedResult.bad("Data is empty");
         }
-        if (!Number.isInteger(data.version) || data.version < 0) {
+        if (!savegameInterfaces[data.version] && (!Number.isInteger(data.version) || data.version < 0)) {
```
Comment (tobspr) [inline src/js/core/read_write_proxy.js:321]: shouldn't this be an `||`? am I missing something here?

File: src/js/game/base_item.js [hunk 9]
```diff
@@ -76,9 +73,15 @@ export class BaseItem extends BasicSerializableObject {
...
+    drawItemCenteredClipped(
+        x,
+        y,
+        parameters,
+        diameter = globalConfig.defaultItemDiameter,
+        background = true
+    ) {
         if (parameters.visibleRect.containsCircle(x, y, diameter / 2)) {
-            this.drawItemCenteredImpl(x, y, parameters, diameter);
+            this.drawItemCenteredImpl(x, y, parameters, diameter, background);
```
Comment (tobspr) [inline src/js/game/base_item.js:84]: what is background meaning here?

File: src/js/game/building_codes.js [hunk 10]
```diff
@@ -56,40 +50,21 @@ export function registerBuildingVariant(
...
-    const result = variantsCache.get(hash);
-    if (G_IS_DEV) {
-        if (!result) {
-            console.warn("Known hashes:", Array.from(variantsCache.keys()));
-            assertAlways(false, "Building not found by data: " + hash);
-        }
-    }
-    return result;
+export function getCodeFromBuildingData(metaBuilding, variant, rotationVariant = 0) {
+    return metaBuilding.getId() + "/" + variant + "/" + rotationVariant;
```
Comment (tobspr) [inline src/js/game/building_codes.js:69]: this is very inefficent, since it is generating a long string. it also bloats up the savegame by a lot. have to find a better solution for this somehow.

Maybe the builtin buildings could have an integer code, the mod buildings a string?

File: src/js/game/buildings/analyzer.js [hunk 11]
```diff
@@ -3,51 +3,120 @@ import { enumDirection, Vector } from "../../core/vector";
...
+    getSilhouetteColor(variant) {
+        return MetaAnalyzerBuilding.silhouetteColors[variant]();
     }
 
     /**
      * @param {GameRoot} root
      */
     getIsUnlocked(root) {
-        return root.hubGoals.isRewardUnlocked(enumHubGoalRewards.reward_virtual_processing);
+        return this.getAvailableVariants(root).length > 0;
```
Comment (tobspr) [inline src/js/game/buildings/analyzer.js:26]: ah yeah, I should have done this much ealier. Maybe we can do it for all buildings like this ?

File: src/js/game/buildings/analyzer.js [hunk 12]
```diff
@@ -68,12 +137,74 @@ export class MetaAnalyzerBuilding extends MetaBuilding {
...
             new LogicGateComponent({
                 type: enumLogicGateType.analyzer,
             })
-        );
-    }
-}
+        ),
+];
+
+MetaAnalyzerBuilding.silhouetteColors = {
```
Comment (tobspr) [inline src/js/game/buildings/analyzer.js:150]: I don't like the mapping for variants here. Why not make a variant its own class?

**Review by tobspr — CHANGES_REQUESTED** [review]
First part of the review

File: src/css/animations.scss (hunk 3 above)
Comment (DJ1TJOO) [reply src/css/animations.scss:16]: I think that was from an older version of shapez.io

File: src/js/core/read_write_proxy.js (hunk 8 above)
Comment (DJ1TJOO) [reply src/js/core/read_write_proxy.js:321]: It also supports string for savegame interfaces and if its not in the savegameInterfaces it doesn't exists. Maybe i need to remove the number check

File: src/js/game/base_item.js (hunk 9 above)
Comment (DJ1TJOO) [reply src/js/game/base_item.js:84]: Draw a background or not for shapes

File: src/js/game/buildings/analyzer.js (hunk 11 above)
Comment (DJ1TJOO) [reply src/js/game/buildings/analyzer.js:26]: Sure

File: electron/index.js (hunk 2 above)
Comment (DJ1TJOO) [reply electron/index.js]: I've done this now

File: src/css/animations.scss (hunk 3 above)
Comment (DJ1TJOO) [reply src/css/animations.scss:16]: Fixed back to normal

Comment (DJ1TJOO) [reply src/css/animations.scss:16]: But missing spaces

File: src/css/dynamic_ui.scss (hunk 4 above)
Comment (DJ1TJOO) [reply src/css/dynamic_ui.scss]: Fixed has missing spaces, needs fixing

File: src/js/application.js (hunk 6 above)
Comment (DJ1TJOO) [reply src/js/application.js:114]: Removed

File: src/js/core/config.local (hunk 7 above)
Comment (DJ1TJOO) [reply src/js/core/config.local]: Removed

File: src/js/game/buildings/analyzer.js (hunk 11 above)
Comment (DJ1TJOO) [reply src/js/game/buildings/analyzer.js:26]: I already is for all buildings, maybe make it default in metabuilding?

File: src/js/game/buildings/analyzer.js (hunk 12 above)
Comment (DJ1TJOO) [reply src/js/game/buildings/analyzer.js:150]: Have done that in https://github.com/DJ1TJOO/shapez.io/tree/modloader-variant-classes. Note this is missing a lot. It's just for feedback right now

File: src/js/game/buildings/analyzer.js (hunk 11 above)
Comment (DJ1TJOO) [reply src/js/game/buildings/analyzer.js]: I made it default in the metabuilding.

File: src/js/core/read_write_proxy.js (hunk 8 above)
Comment (DJ1TJOO) [reply src/js/core/read_write_proxy.js:321]: Removed number check

File: README.md (hunk 1 above)
Comment (DJ1TJOO) [reply README.md:136]: It has loads of slashes because I need to escape the *****, **_**  and **\**. This is not shown on the github page. I now also changed it to a heading and steps

File: src/css/animations.scss (hunk 3 above)
Comment (DJ1TJOO) [reply src/css/animations.scss]: Fully fixed

File: src/css/dynamic_ui.scss (hunk 4 above)
Comment (DJ1TJOO) [reply src/css/dynamic_ui.scss]: Fixed

File: src/css/dynamic_ui.scss (hunk 5 above)
Comment (DJ1TJOO) [reply src/css/dynamic_ui.scss]: Fixed

File: src/js/game/buildings/analyzer.js (hunk 12 above)
Comment (DJ1TJOO) [reply src/js/game/buildings/analyzer.js:143]: Also fixed circular dependencie for **defaultVariantBuilding**

File: src/js/game/building_codes.js (hunk 10 above)
Comment (DJ1TJOO) [reply src/js/game/building_codes.js:69]: It can all be a number. The problem is that when you load a world with different mods different buildings are registered to the numbers

Comment (DJ1TJOO) [reply src/js/game/building_codes.js:69]: For 10.000 buildings it's about 0.23mb difference

Comment (DJ1TJOO) [reply src/js/game/building_codes.js:69]: I tested with some worlds:
One went from 11.4mb to 11.5mb
And another went from 3.70mb to 3.71mb

File: README.md (hunk 1 above)
Comment (tobspr) [reply README.md]: Why don't you use code formatting?

File: README.md [hunk 13]
```diff
@@ -130,3 +130,9 @@ For most assets I use Adobe Photoshop, you can find them <a href="//github.com/t
 All assets will be automatically rebuilt into the atlas once changed (Thanks to dengr1065!)
 
 <img src="https://i.imgur.com/W25Fkl0.png" alt="shapez.io Screenshot">
+
+### Generating documentation:
+
+1. Replace name.replace('\*', '\\\\\*') with name.replace('\*', '\\\\\*').replace('[', '\\\\[').replace(']', '\\\\]').replace('(', '\\\\(').replace(')', '\\\\)') in node_modules\esdoc-publish-html-plugin\out\src\Builder\DocBuilder.js:116:20)
```
Comment (tobspr) [inline README.md:136]: Just use backticks here :)

File: electron/index.js [hunk 14]
```diff
@@ -153,61 +155,80 @@ ipcMain.on("exit-app", (event, flag) => {
 });
 
 function performFsJob(job) {
-    const fname = path.join(storePath, job.filename);
+    let fname = path.join(storePath, job.filename);
+    if (job.mods) fname = path.join(modsPath, job.filename);
```
Comment (tobspr) [inline electron/index.js:159]: wouldn't this allow passing `../../../something.txt` as a filename?

File: README.md (hunk 1 above)
Comment (DJ1TJOO) [reply README.md]: What do you mean?

File: src/js/application.js [hunk 15]
```diff
@@ -413,3 +396,18 @@ export class Application {
         this.checkResize(true);
     }
 }
+
+Application.trackClicks = undefined;
```
Comment (tobspr) [inline src/js/application.js:400]: Why not add it as a static member?

```js
class Application {
  static trackClicks = undefined;

}
```

File: src/js/game/base_item.js (hunk 9 above)
Comment (tobspr) [reply src/js/game/base_item.js:84]: what's the use case? which background?

File: src/js/game/building_codes.js (hunk 10 above)
Comment (tobspr) [reply src/js/game/building_codes.js:69]: is that the compressed version?

File: src/js/game/buildings/analyzer.js [hunk 16]
```diff
@@ -3,51 +3,113 @@ import { enumDirection, Vector } from "../../core/vector";
...
+     * @param {Entity} entity
+     * @param {number} rotationVariant
+     * @param {string} variant
+     */
+    updateVariants(entity, rotationVariant, variant) {
+        MetaAnalyzerBuilding.componentVariations[variant](entity, rotationVariant);
+    }
+}
+
+MetaAnalyzerBuilding.setupEntityComponents = [
```
Comment (tobspr) [inline src/js/game/buildings/analyzer.js:111]: same comments as above, this can simply be a static member

**Review by tobspr — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (tobspr) [reply README.md]: https://docs.github.com/en/github/writing-on-github/basic-writing-and-formatting-syntax#quoting-code

File: src/js/game/buildings/analyzer.js (hunk 12 above)
Comment (tobspr) [reply src/js/game/buildings/analyzer.js:143]: which files should I check there?

File: src/js/game/building_codes.js (hunk 10 above)
Comment (DJ1TJOO) [reply src/js/game/building_codes.js:69]: The bin files

File: src/js/game/buildings/analyzer.js (hunk 12 above)
Comment (DJ1TJOO) [reply src/js/game/buildings/analyzer.js:143]: https://github.com/DJ1TJOO/shapez.io/blob/modloader-variant-classes/src/js/game/meta_building.js
https://github.com/DJ1TJOO/shapez.io/blob/modloader-variant-classes/src/js/game/meta_building_variant.js
https://github.com/DJ1TJOO/shapez.io/blob/modloader-variant-classes/src/js/game/meta_building_registry.js

Examples:
https://github.com/DJ1TJOO/shapez.io/tree/modloader-variant-classes/src/js/game/buildings/variants
https://github.com/DJ1TJOO/shapez.io/blob/modloader-variant-classes/src/js/game/buildings/analyzer.js
https://github.com/DJ1TJOO/shapez.io/blob/modloader-variant-classes/src/js/game/buildings/balancer.js

File: README.md (hunk 1 above)
Comment (DJ1TJOO) [reply README.md]: Done that. Forgot readme has that

File: README.md (hunk 13 above)
Comment (DJ1TJOO) [reply README.md]: Done that

File: electron/index.js (hunk 14 above)
Comment (DJ1TJOO) [reply electron/index.js:159]: Yes forgot that

File: src/js/application.js (hunk 15 above)
Comment (DJ1TJOO) [reply src/js/application.js:400]: Going to do that, I'm also going to do this in the variant-class branch

File: electron/index.js (hunk 2 above)
Comment (tobspr) [reply electron/index.js]: > maybe we can somehow ensure that the game can not read any file outside of its game folder?

File: src/js/game/building_codes.js (hunk 10 above)
Comment (tobspr) [reply src/js/game/building_codes.js:69]: the bin files are 11 mb?

File: src/js/game/buildings/analyzer.js (hunk 12 above)
Comment (tobspr) [reply src/js/game/buildings/analyzer.js:143]: The change itself looks nice, although very breaking. Let's start with the modloader only

File: src/js/application.js (hunk 15 above)
Comment (tobspr) [reply src/js/application.js:400]: Alright!

File: src/js/game/base_item.js (hunk 9 above)
Comment (DJ1TJOO) [reply src/js/game/base_item.js:84]: The shadow of the items

File: src/js/game/building_codes.js (hunk 10 above)
Comment (DJ1TJOO) [reply src/js/game/building_codes.js:69]: Yes and it has about 200000 buildings

File: electron/index.js (hunk 14 above)
Comment (DJ1TJOO) [reply electron/index.js]: Fixed this
